File: 프로그래머스/3/43163. 단어 변환/단어 변환.js
var min = 0;
var isUsed;

function solution(begin, target, words) {
    
    var n = words.length;
    isUsed = Array(n).fill(false);
    min = n;
    
    dfs(begin, target, words, 0);
    
    if(min == n) min = 0;

    return min;
}

function dfs(begin, target, words, change){
    
    if(change >= min) return; 
    
    if(begin == target){
        min = Math.min(min, change);
        return;
    }
    
    for(let i=0; i<words.length; i++){
        var now = words[i];
        if(isUsed[i]) continue;
        var cnt = 0;
        for(let j=0; j<now.length; j++){
            if(begin.charAt(j) != now.charAt(j)) cnt++;
        }
        
        if(cnt == 1){
            isUsed[i] = true;
            dfs(now, target, words, change+1);
            isUsed[i] = false;
        }
    }
}

File: 프로그래머스/3/43163. 단어 변환/단어 변환.java
import java.util.*;

class Solution {
    
    static int min = 0;
    static boolean isUsed[];
    
    public int solution(String begin, String target, String[] words) {
        
        int n = words.length;
        min = n;
        isUsed = new boolean[n+1];
        
        dfs(begin, target, words, 0);
        
        if(min == words.length) min = 0;
        
        return min;
    }
    
    public void dfs(String begin, String target, String[] words, int change){
        
        if(change >= min) return;
        
        if(begin.equals(target)){
            min = Math.min(change, min);
            return;
        }
        
        for(int i=0; i<words.length; i++){
            String now = words[i];
            if(isUsed[i]) continue;
            int cnt = 0;
            for(int j=0; j<now.length(); j++){
                if(begin.charAt(j) != now.charAt(j)) cnt++;
            }
            
            if(cnt == 1){
                isUsed[i] = true;
                dfs(now, target, words, change+1);
                isUsed[i] = false;
            }
        }
        
    }
       
}